features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp

```diff
@@ -0,0 +1,1234 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ble/SecurityManager.h"
+#include "ble/pal/PalSecurityManager.h"
+#include "ble/generic/GenericSecurityManager.h"
+#if defined(MBEDTLS_CMAC_C)
+#include "mbedtls/cmac.h"
+#endif
+
+namespace ble {
+namespace generic {
+
+/* Implements SecurityManager */
+
+////////////////////////////////////////////////////////////////////////////
+// SM lifecycle management
+//
+
+ble_error_t GenericSecurityManager::init(
+    bool bondable,
+    bool mitm,
+    SecurityIOCapabilities_t iocaps,
+    const Passkey_t passkey,
+    bool signing
+) {
+    _db.restore();
+    _pal.set_io_capability((io_capability_t::type) iocaps);
+    _pal.set_display_passkey(PasskeyAscii::to_num(passkey));
+    _legacy_pairing_allowed = true;
+
+    bool secure_connections;
+    _pal.get_secure_connections_support(secure_connections);
+
+    _default_authentication.set_bondable(bondable);
+    _default_authentication.set_mitm(mitm);
+    _default_authentication.set_secure_connections(secure_connections);
+    _default_authentication.set_keypress_notification(true);
+
+    _default_key_distribution.set_signing(signing);
+    if (signing) {
+        init_signing();
+    }
+
+    _connection_monitor.set_connection_event_handler(this);
+    _pal.set_event_handler(this);
+
+    _pal.generate_public_key();
+
+    return BLE_ERROR_NONE;
+}
+
+ble_error_t GenericSecurityManager::reset(void) {
+    _db.sync();
+    _public_keys_generated = false;
+    SecurityManager::reset();
+
+    return BLE_ERROR_NONE;
+}
+
+ble_error_t GenericSecurityManager::preserveBondingStateOnReset(bool enabled) {
+    _db.set_restore(enabled);
+    return BLE_ERROR_NONE;
+}
+
+////////////////////////////////////////////////////////////////////////////
+// List management
+//
+
+ble_error_t GenericSecurityManager::purgeAllBondingState(void) {
+    _db.clear_entries();
+    return BLE_ERROR_NONE;
+}
+
+ble_error_t GenericSecurityManager::generateWhitelistFromBondTable(Gap::Whitelist_t *whitelist) const {
+    if (eventHandler) {
+        _db.generate_whitelist_from_bond_table(
+            mbed::callback(eventHandler, &::SecurityManager::SecurityManagerEventHandler::whitelistFromBondTable),
+            whitelist
+        );
+    }
+    return BLE_ERROR_NONE;
+}
+
+////////////////////////////////////////////////////////////////////////////
+// Pairing
+//
+
+ble_error_t GenericSecurityManager::requestPairing(connection_handle_t connection) {
+    ControlBlock_t *cb = get_control_block(connection);
+    if (!cb) {
+        return BLE_ERROR_INVALID_PARAM;
+    }
+
+    if (!_legacy_pairing_allowed && !_default_authentication.get_secure_connections()) {
+        return BLE_ERROR_OPERATION_NOT_PERMITTED;
```

Comment:
Is it not permitted or not possible?

Reply:
I have a limited selection of errors, there is no: logically impossible

Reply:
What do you think of `BLE_ERROR_INVALID_STATE` ?